File: Project/src/scripts/algorithm.js
var mongodb = require('mongodb');

function getPossibles(values, minSum, maxSum) {
    var possibles = [];
    values.sort(function(a, b) { return a[1] - b[1]; });
    for (i = 0; i < values.length; i++) {
        var minSumTwo = minSum - values[i][1];
        var maxSumTwo = maxSum - values[i][1];
        var j = i + 1;
        var k1 = values.length - 1; 
        var k2 = values.length - 1;
        while (k2 > j) {
            while (values[j][1] + values[k2][1] > maxSumTwo && k2 > j) {
                k2--;
            }
            for (k = k1 + 1; k < k2 + 1; k++) {
                possibles.push([values[i][0], values[j][0], values[k][0]]);
            }
            k1 = Math.min(k1, k2);
            while (values[j][1] + values[k1][1] >= minSumTwo && k1 > j) {
                possibles.push([values[i][0], values[j][0], values[k1][0]]);
                k1--;
            };
            j++;
            k1 = Math.max(k1, j);
        }
    }
    return possibles;
}

function getAllRecipes(db, callback) {
    var recipes = [];
    var cursor = db.collection('recipes').find();
    cursor.forEach(function (doc) {
        recipes.push(doc['id']);
    }, function (err) {
        callback(recipes);
    });
}

function getFavorites(db, userID, callback) {
    var favorites = [];
    var cursor = db.collection('favourite').find({'userID': userID});
    cursor.forEach(function (doc) {
        favorites.push(doc['recipeID']);
    }, function (err) {
        callback(favorites);
    });
}

function getClusters(db, callback) {
    var clusters = {};
    var cursor = db.collection('clustering').find();
    cursor.forEach(function (doc) {
        clusters[doc['recipe']] = doc['cluster'];
    }, function (err) {
        callback(clusters);
    });
}

function getScores(db, recipes, favorites, callback) {
    var scores = {};
    for (i in recipes) {
        scores[recipes[i]] = 0;
    }
    // var favoritesSet = new Set(favorites);
    var favoritesSet = {};
    for (i in favorites) {
        favoritesSet[favorites[i]] = true;
    }

    var cursor = db.collection('similarities').find();
    cursor.forEach(function (doc) {
        recipeA = doc['recipe_A'];
        recipeB = doc['recipe_B'];
        similarity = doc['similarity'];
        if (recipeA in scores && recipeB in scores) {
            // if (similarity > scores[recipeA] && favoritesSet.has(recipeB)) {
            if (similarity > scores[recipeA] && recipeB in favoritesSet) {
                scores[recipeA] = similarity;
            }
            // if (similarity > scores[recipeB] && favoritesSet.has(recipeA)) {
            if (similarity > scores[recipeB] && recipeA in favoritesSet) {
                scores[recipeB] = similarity;
            }
        }
    }, function (err) {
        callback(scores);
    });
}

function getNutrition(db, element, recipes, callback) {
    var cursor = db.collection('nutrition').find({
                                     'attribute': element,
                                     'recipeID': {'$in': recipes}})
    results = [];
    cursor.forEach(function (doc) {
        results.push([doc['recipeID'], doc['value']]);
    }, function (err) {
        callback(results);
    });
}

function suggestRecipes(db, userID, numSuggestions, callback) {
    getAllRecipes(db, function (recipes) {
        getFavorites(db, userID, function (favorites) {
            getScores(db, recipes, favorites, function(scores) {
                recipes.sort(function(a, b) { return scores[b]-scores[a]; });
                var topRecipes = recipes.slice(0,1000);
                getNutrition(db, 'ENERC_KCAL', topRecipes, function (calories) {
                    // TODO: get actual calorie cutoffs form database and user info
                    var triples = getPossibles(calories, 1950, 2000);
                    getClusters(db, function (clusters) {
                        var dayScores = triples.map(function (x) {
                            var sum = x.map(function (r) {
                                return scores[r];
                            }).reduce(function (a,b) {
                                return a + b;
                            });
                            return [x, sum];
                        });
                        dayScores.sort(function (a,b) {
                            return b[1] - a[1];
                        });

                        // var takenClusters = new Set();
                        var takenClusters = {};
                        var weekRecipes = [];
                        var i = 0;
                        while (weekRecipes.length < numSuggestions && i < dayScores.length) {
                            var candidate = dayScores[i][0];
                            var candidateClusters = candidate.map(function (x) {
                                return clusters[x];
                            });

                            var duplicateExists = false;
                            candidateClusters.sort();
                            for (j = 1; j < candidateClusters.length; j++) {
                                if (candidateClusters[j] == candidateClusters[j-1]) {
                                    duplicateExists = true;
                                }
                            }

                            if (!duplicateExists) {
                                if (!candidateClusters.some(function (x) {
                                    return x in takenClusters;
                                })) {
                                    weekRecipes.push(candidate);
                                    for (j in candidateClusters) {
                                        takenClusters[candidateClusters[j]] = true;
                                    }
                                }
                            }
                            i++;
                        }

                        callback(weekRecipes);
                    });
                });
            });
        });
    });
}

function getRecipeSuggestions(userID, numSuggestions, callback) {
    var url = 'mongodb://localhost/users';
    mongodb.MongoClient.connect(url, function(err, db) {
        suggestRecipes(db, userID, numSuggestions, function (suggestions) {
            db.close();
            callback(suggestions);
        });
    });
}

exports.getPossibles = getPossibles;
exports.getAllRecipes = getAllRecipes;
exports.getFavorites = getFavorites;
exports.getClusters = getClusters;
exports.getScores = getScores;
exports.getNutrition = getNutrition;
exports.suggestRecipes = suggestRecipes;
exports.getRecipeSuggestions = getRecipeSuggestions;

// getRecipeSuggestions(console.log);

File: Project/src/scripts/algorithm.py
import pymongo

def getAllRecipes(db):
    recipes = []
    cursor = db.recipes.find()
    for recipe in cursor:
        recipes.append(recipe['id'])
    return recipes

def getFavorites(db):
    # TODO: favorites in the database
    return ['Slow-Cooker-Chinese-Three-Cup-Chicken-1335215',
            'Chinese-Barbecued-Pork-_Char-Siu_-622609',
            'Orange-Chicken-1327985',
            'Har-Gow-1324819',
            'Ground-Beef-Bulgogi-1286788',
            'PotStickers-1329410',
            'Shrimp-Wonton-Soup-1315067',
            'Sesame-Soba-Noodles-623025',
            'Quick-and-Easy-Bacon-Fried-Rice-1300204',
            'Asian-Burgers-with-Spicy-Coleslaw-1332436',
            'Creamy-Mushroom-Soup-1217212',
            'Quick-and-Easy-Egg-Drop-Soup-1019556',
            'Chicken-Pho-1036448',
            'Asian-Chicken-Noodle-Soup-1300938']

def getClusters(db):
    clusters = {}
    for clusL in db.clustering.find():
        clusters[clusL['recipe']] = clusL['cluster']
    return clusters

def getScores(db, recipes, favorites):
    scores = {}
    for recipe in recipes:
        scores[recipe] = 0

    for simL in db.similarities.find():
        recipeA = simL['recipe_A']
        recipeB = simL['recipe_B']
        similarity = simL['similarity']
        if recipeA in scores and recipeB in scores:
            if similarity > scores[recipeA] and recipeB in favorites:
                scores[recipeA] = similarity
            if similarity > scores[recipeB] and recipeA in favorites:
                scores[recipeB] = similarity
    return scores

def getRequirements(db):
    pass

def getNutrition(db, element, recipes):
    results = []
    for res in db.nutrition.find({'attribute': element,
                                  'recipeID': {'$in': recipes}}):
        results.append((res['recipeID'], res['value']))
    return results

def getPossibles(values, minSum, maxSum):
    possibles = []
    values.sort(key=lambda value: value[1])
    for i in range(len(values)):
        minSumTwo = minSum - values[i][1]
        maxSumTwo = maxSum - values[i][1]
        j = i+1
        k1 = k2 = len(values)-1
        while k2 > j:
            while values[j][1] + values[k2][1] > maxSumTwo and k2 > j:
                k2 -= 1
            for k in range(k1+1,k2+1):
                possibles.append((values[i][0], values[j][0], values[k][0]))
            k1 = min(k1, k2)
            while values[j][1] + values[k1][1] >= minSumTwo and k1 > j:
                possibles.append((values[i][0], values[j][0], values[k1][0]))
                k1 -= 1
            j += 1
            k1 = max(k1, j)
    return possibles

def suggestRecipes(db):
    recipes = getAllRecipes(db)

    favorites = getFavorites(db)
    scores = getScores(db, recipes, favorites)
    topRecipes = [recipe[0] for recipe in sorted(scores.iteritems(),
                                                 key=lambda x: x[1],
                                                 reverse=True)[:1000]]

    calories = getNutrition(db, 'ENERC_KCAL', topRecipes)
    # TODO: get actual calorie values from database and user info
    triples = getPossibles(calories, 1950, 2000)

    clusters = getClusters(db)
    dayScores = map(lambda x: (x, sum([scores[i] for i in x])), triples)
    dayScores.sort(key=lambda x: x[1], reverse=True)

    takenClusters = set([])
    weekRecipes = []
    i = 0
    while len(weekRecipes) < 7 and i < len(dayScores):
        candidate = dayScores[i][0]
        candidateClusters = map(lambda x: clusters[x] if x in clusters
                                                      else 0, candidate)
        if not takenClusters.intersection(candidateClusters):
            weekRecipes.append(candidate)
            takenClusters.update(candidateClusters)
        i += 1

    return weekRecipes

def main():
    # TODO: get userid
    client = pymongo.MongoClient()
    db = client.users
    suggestRecipes(db)


if __name__ == '__main__':
    main()
